src/tracing.rs

```diff
@@ -0,0 +1,366 @@
+use crate::{
+    callframe::{Callframe, NearCallFrame},
+    decommit::is_kernel,
+    predication::{self, Predicate},
+    VirtualMachine,
+};
+use eravm_stable_interface::*;
+use std::cmp::Ordering;
+
+impl<T> StateInterface for VirtualMachine<T> {
+    fn read_register(&self, register: u8) -> (u256::U256, bool) {
+        (
+            self.state.registers[register as usize],
+            self.state.register_pointer_flags & (1 << register) != 0,
+        )
+    }
+
+    fn set_register(&mut self, register: u8, value: u256::U256, is_pointer: bool) {
+        self.state.registers[register as usize] = value;
+
+        self.state.register_pointer_flags &= !(1 << register);
+        self.state.register_pointer_flags |= u16::from(is_pointer) << register;
+    }
+
+    fn number_of_callframes(&self) -> usize {
+        self.state
+            .previous_frames
+            .iter()
+            .map(|frame| frame.near_calls.len() + 1)
+            .sum::<usize>()
+            + self.state.current_frame.near_calls.len()
+            + 1
+    }
+
+    fn callframe(&mut self, n: usize) -> impl CallframeInterface + '_ {
```

Comment:
Another dumb question: What does `n` mean? At first I thought it means an index of far and near call frames considered together, but this doesn't look to be the case; `n` isn't modified when iterating over far call frames, while it would need to be decreased in the `Ordering::Greater` branch of the `match` below. Conceptually, I think it could make sense to treating far and near call frames differently in the VM state interface, since there's a clear semantic distinction between them. This would allow simplifying accessing methods.

Reply:
Yes, this code is buggy and I should write a test for it.

I to treat near and far frames as the same thing because what is stored in near and what in far only is pretty vague. This way the API doesn't break if near calls get the ability to mutate some new field.